### PR TITLE
Add queue for referral code creation events

### PR DESCRIPTION
**Pull request by tomrf1**
Braze cannot be trusted to rate-limit its referral code creation requests.
In extreme cases it has used up all db connections, preventing other services from using the db.
The solution is to queue the requests.

Current architecture:
`braze --> API gateway --> lambda --> DB`

New architecture:
`braze --> API gateway --> sqs --> lambda --> DB`

To protect the database, we can control the throughput with:
1. sqs batch size
2. lambda concurrency

We tested batch size of 10 and concurrency of 10 in CODE and there were no failures, and the DB did not struggle.
The PROD RDS instance size is larger, so this configuration should be fine.
We tested this using the gatling script in this repo, and the lambda processed about 5.5k events in 1 minute.

### Implementation details

- Add a new SQS queue
- Create a custom api gateway, instead of using the implicit `ServerlessRestApi` one that is generated by SAM.
- Integrate new `/referral-code-queue` api gateway endpoint with the queue
- Re-create the existing `/referral-count` GET endpoint (unfortunately there is now more boilerplate for this part, because it's using the custom api gateway)
- Integrate the create-code lambda with SQS, rather than api gateway
- Change the create-code lambda logic to recieve a batch of sqs events.
- Add `ON CONFLICT DO NOTHING` to the `INSERT` request, because we do not want the lambda to fail if it processes the same event twice. This is possible if e.g. a batch partially fails, and we subsequently have to retry events from that batch that have already succeeded.

TODO -
The `/referral-count` also hits the DB. Should we have a max concurrency for this lambda as well?

### RDS connections (CODE)
Screen Shot 2020-10-26 at 14 32 39
### RDS CPU (CODE)
Screen Shot 2020-10-26 at 14 32 56

### CODE REVIEW

File: cfn.yaml [hunk 1]
```diff
@@ -51,6 +51,99 @@ Globals:
...
+        - PolicyName: ApiQueuePolicy
+          PolicyDocument:
+            Version: '2012-10-17'
+            Statement:
+              - Effect: Allow
+                Action:
+                  - sqs:ReceiveMessage
+                  - sqs:SendMessage
+                Resource: !GetAtt ReferralCodeQueue.Arn
+        - PolicyName: LambdaInvokePolicy
```
Comment (tomrf1) [inline cfn.yaml:81]: give the api gateway permission to invoke the referral-count lambda

File: cfn.yaml [hunk 2]
```diff
@@ -51,6 +51,99 @@ Globals:
...
+                  schema:
+                    $ref: "#/definitions/Empty"
+              x-amazon-apigateway-integration:
+                credentials: !GetAtt ApiGatewayRole.Arn
+                uri: !Sub "arn:aws:apigateway:${AWS::Region}:sqs:path//"
+                responses:
+                  default:
+                    statusCode: "200"
+                    responseTemplates:
+                      "application/json": "#set($inputRoot = $input.path('$'))"
```
Comment (tomrf1) [inline cfn.yaml:133]: tell api gateway to throw away the response from sqs, as it is not helpful to the client. Just a 200 will do

File: cfn.yaml [hunk 3]
```diff
@@ -51,6 +51,99 @@ Globals:
...
+          "/referral-count":
+            "get":
+              responses:
+                "200":
+                  description: "200 response"
+                  schema:
+                    $ref: "#/definitions/Empty"
+              x-amazon-apigateway-integration:
+                credentials: !GetAtt ApiGatewayRole.Arn
+                httpMethod: "POST"
```
Comment (tomrf1) [inline cfn.yaml:108]: here `httpMethod` is for the backend (lambda) integration, hence POST. Not to be confused with the method of the api gateway endpoint, which is GET!

File: cfn.yaml (hunk 2 above)
Comment (paulbrown1982) [reply cfn.yaml:133]: I didn't know API Gateway lets you code Velocity templates in CloudFormation!

File: src/create-code/gatling-test/src/test/scala/BasicSimulation.scala [hunk 4]
```diff
@@ -10,9 +10,10 @@ class BrazeReferralTest extends Simulation {
...
 
   def requestBody: String = s"""{
-											|		"code": "$randomNumber",
-											|		"email": "$randomNumber@gu.com",
-											|		"source": "GATLING"
+											|		"code": "gatlingTestCode",
+											|		"email": "gatling.test@gu.com",
+											|		"source": "GATLING",
+                      |   "campaignId": "gatling-test-campaign"
 											|}""".stripMargin
```
Comment (paulbrown1982) [inline src/create-code/gatling-test/src/test/scala/BasicSimulation.scala:17]: Can the whitespace here be tidied up?

File: cfn.yaml (hunk 2 above)
Comment (tomrf1) [reply cfn.yaml:133]: This is velocity inside swagger inside cloudformation yaml
It's not beautiful

File: src/create-code/gatling-test/src/test/scala/BasicSimulation.scala (hunk 4 above)
Comment (tomrf1) [reply src/create-code/gatling-test/src/test/scala/BasicSimulation.scala:17]: just realised this file has tabs in it!

**Pull request merged**